Topics/System_Design/Intro_To_Microservices/Intro_To_Microservices.md

```diff
@@ -0,0 +1,165 @@
+
+In summary, while microservices architecture offers numerous advantages, CSC301 students should carefully consider the associated challenges before adopting this architectural style for software development projects. Understanding the benefits and drawbacks of microservices architecture is essential for making informed architectural decisions in alignment with the principles taught in CSC301.
+
+## Principles and Characteristics of Microservices Architecture <a name="principles"></a>
+
+In the realm of CSC301H1: Introduction to Software Engineering, understanding the principles and characteristics of modern software architectures is essential. Microservices architecture, a central concept in contemporary software development, aligns closely with the principles taught in this course.
+
+1. **Modularity**: Microservices architecture emphasizes modularity, breaking down complex applications into smaller, self-contained services—a concept akin to the modular design principles discussed in CSC301. This modularity promotes maintainability and enables teams to work on services independently, fostering agile development practices.
+
+2. **Autonomy**: The autonomy of microservices resonates with the agile development methods taught in CSC301. Each service operates independently, allowing teams to work on services concurrently and deploy changes without impacting other parts of the system. This autonomy promotes rapid iteration and continuous deployment, core tenets of agile development.
+
+3. **Loose Coupling**: Microservices architecture promotes loose coupling between services, facilitating flexibility and enabling teams to adapt to changing requirements—an essential aspect of agile development. By interacting through well-defined APIs, services remain decoupled, ensuring that changes in one service do not ripple through the entire architecture.
+
+4. **Scalability**: The scalability of microservices architecture aligns with discussions on scalability and resource management in CSC301. Individual services can be scaled independently based on demand, optimizing resource utilization and ensuring optimal performance—a fundamental aspect of designing scalable software systems.
+
+5. **Resilience**: Microservices architecture promotes resilience by isolating failures within individual services, a concept reminiscent of fault tolerance discussions in CSC301. If one service encounters an issue, it does not propagate to other parts of the system, enhancing system stability and reliability.
+
+6. **Polyglot Persistence**: The flexibility of microservices architecture to support polyglot persistence reflects discussions on database design and management in CSC301. Each service can choose the most suitable database technology based on its requirements, optimizing data storage and retrieval strategies.
```

Comment:
What is Polyglot? We may have discussed this during class, but I can't recall. One line definition would've been helpful

Reply:
It was only briefly touched upon during the class, and this pointer goes slightly overboard in terms of concepts, but I have included it anyways because of its significance.